zkvm/src/predicate.rs

```diff
@@ -181,75 +156,221 @@ impl Into<CompressedRistretto> for Predicate {
+            p.compress()
+        };
 
-    fn bytecode(prog: &Program) -> Vec<u8> {
-        let mut prog_vec = Vec::new();
-        prog.encode(&mut prog_vec);
-        prog_vec
+        Ok(Self {
+            leaves,
+            blinding_key,
+            key,
+            precomputed_point,
+            adjustment_factor,
+        })
     }
 
-    #[test]
-    fn valid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let pred = Predicate::unblinded_program(prog.clone());
-        let op = pred.prove_program_predicate(&bytecode(&prog), &[]);
-        assert!(op.verify().is_ok());
+    /// Returns the adjustment factor for signing
+    // TODO: Instead, we would rather return a "key witness" object like musig::Multikey.
+    // That would directly store the adjustment factor.
+    pub fn adjustment_factor(&self) -> Scalar {
+        self.adjustment_factor
     }
 
-    #[test]
-    fn invalid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let prog2 = Program::build(|p| p.dup(1));
-        let pred = Predicate::unblinded_program(prog);
-        let op = pred.prove_program_predicate(&bytecode(&prog2), &[]);
-        assert!(op.verify().is_err());
+    /// Creates the call proof and returns that with the program at an index.
+    pub fn create_callproof_program(
+        &self,
+        prog_index: usize,
+    ) -> Result<(CallProof, Program), VMError> {
+        let possible_leaf = self.get_leaf(2 * prog_index);
+        let leaf_index = match possible_leaf {
+            PredicateLeaf::Blinding(_) => 2 * prog_index + 1,
+            PredicateLeaf::Program(_) => 2 * prog_index,
+        };
+        let leaves = self.get_leaves();
+        let tree = MerkleTree::build(b"ZkVM.taproot", leaves);
+        let neighbors = tree.create_path(leaf_index)?;
+        let verification_key = self.get_key().clone();
+        let call_proof = CallProof {
+            verification_key,
+            neighbors,
+        };
+        let program = self.get_leaf(leaf_index).clone().to_program()?;
+        Ok((call_proof, program))
     }
 
-    #[test]
-    fn valid_disjunction1() {
-        let gens = PedersenGens::default();
+    fn create_merkle_leaves(progs: &Vec<Program>, blinding_key: [u8; 32]) -> Vec<PredicateLeaf> {
+        let mut t = Transcript::new(b"ZkVM.taproot-derive-blinding");
+        let n: u64 = progs.len() as u64;
+        t.commit_u64(b"n", n);
+        t.commit_bytes(b"key", &blinding_key);
+        for prog in progs.iter() {
+            let mut buf = Vec::new();
+            prog.encode(&mut buf);
+            t.commit_bytes(b"prog", &buf);
+        }
 
-        // dummy predicates
-        let preds = vec![
-            Predicate::Opaque(gens.B.compress()),
-            Predicate::Opaque(gens.B_blinding.compress()),
-        ];
+        let mut leaves = Vec::new();
+        for prog in progs.iter() {
+            let blinding = rand::thread_rng().gen::<[u8; 32]>();
```

Comment:
the role of the `blinding_key` is to construct the entire tree deterministically.

```suggestion
            let blinding = [0u8;32]; 
            t.challenge_bytes("blinding", &mut blinding);
```